fix(Msal-Angular): Dynamic interceptor authority (#3281)
AzureAD/microsoft-authentication-library-for-js: lib/msal-angular/docs/multi-tenant.md

    });
  }
```

## Multi-tenant access tokens

By default the MsalInterceptor will retrieve access tokens from the users' primary tenant. You can however override this behavior by setting a dynamic authority on the MsalInterceptorConfig.

```js
export function MSALInterceptorConfigFactory(): MsalInterceptorConfiguration {
  const protectedResourceMap = new Map<string, Array<string>>();
  protectedResourceMap.set("https://graph.microsoft.com/v1.0/me", ["user.read"]);
  
  return {
    interactionType: InteractionType.Popup,
    protectedResourceMap,
    authRequest: {
      authority: 'https://login.microsoftonline.com/organizations',
    },

>     dynamicAuthority: (account: AccountInfo) => `https://login.microsoftonline.com/${account.tenantId ?? 'organizations'}` 

Reviewer: Instead of just changing the authority, we can maybe consider being able to change the entire `authRequest`? This would mean changing `authRequest` to support an object literal or a function (that takes in the account, like your PR shows) that returns an object literal. That would help address other scenarios for other properties on the `authRequest`.

Reviewer: Also, if we can make a similar change for MSAL Guard, I think that would beneficial.

Author: I did consider this, but came to the conclusion that you cannot change other properties of the request. The scope is already passed in and changing the redirect uri at runtime is weird since it requires registration at the azure portal.

Also you cannot change any properties on the guard, because either you have an account and the guard just says "ok go ahead" or you don't have an account and then there is no account to change it dynamically.

Reviewer: Any of the properties for the declared types (`Omit<PopupRequest, "scopes"> | Omit<RedirectRequest, "scopes"> | Omit<SilentRequest, "scopes">;`) are changable.

Reviewer: (We may consider adding `account` to teh omit list, cc: @jo-arroyo)

Reviewer: For example:

```js
let authRequest;
if (typeof this.msalInterceptorConfig.authRequest === "function") {
  authRequest = this.msalInterceptorConfig.authRequest(account);
} else {
  authRequest = this.msalInterceptorConfig.authRequest;
}
```

Author: I know those properties can be changed because of the type definition, but I was wondering what properties would be reasonable to change.

And omitting the account might be a good idea 😉

Reviewer: > I know those properties can be changed because of the type definition, but I was wondering what properties would be reasonable to change.

We designed it to allow any of the properties (other than `scopes`) to be set via `authRequest`, and I don't see any reason not to allow of those same properties to be set via a dynamic `authRequest`.

Reviewer: In fact, a few of them (e.g. correlation id) may only be useful if they can be set dynamically.

Reviewer: > And omitting the account might be a good idea

I would actually wait on that. Right now we support the active account or the first account from `getAllAccounts` array, but we may apps to have different logic.

Author: So the logic can be something like `dynamicAuthRequest: (originalAuthRequest: ..., httpReq: HttpEvent<any>): AuthRequest`

Then you can support overriding all properties you want. And you're right it does make sense to set the CorrelationId dynamically.

Reviewer: I would personally prefer the `authRequest` property support two types (typed object or function that returns a typed object), instead of having two separate fields, as I think that will be simpler for us and applications.

@jo-arroyo @tnorling Thoughts?

Reviewer: And as mentioned, I think we should have this change for both the guard and interceptor, given they have the same considerations around building the auth request.

Author: @jasonnutter it also makes sense to still have defaults defined. By having two types on the authRequest configuration you cannot use both at the same time. And have to use dynamic for everything.

Reviewer: > By having two types on the authRequest configuration you cannot use both at the same time.

True, but if you need to set static and dynamic fields, both can be done in the dynamic version, e.g.

```js
authRequest: (httpReq, msalInstance) => {
  return {
     correlation: dyanmicCorrelationId,
     forceRefesh: true
  };
};
```

Reviewer: > it also makes sense to still have defaults defined

Can you please clarify what you mean? The scope and account are the only properties that we set, we don't provide any additional defaults (aside from the defaults set internally by MSAL for acquireTokenSilent, etc).

Reviewer 2: I like the idea of having one `authRequest` property that can either be static or a function. I can forsee lots of confusion about which is used, which should be used, what the difference is, etc. if we keep them separate.

Author: I changed it to be one property that is either a function or a static. And added the same functionality for the MsalGuard

    });
  }
```

> ## Multi-tenant access tokens

Reviewer: I think it would be good to include two sub sections here, one for the guard, and one for the interceptor.

    });
  }
```

## Multi-tenant access tokens

By default the MsalInterceptor will retrieve access tokens from the users' primary tenant. You can however override this behavior by setting a dynamic auth request on the MsalInterceptorConfig.

```js
export function MSALInterceptorConfigFactory(): MsalInterceptorConfiguration {
  const protectedResourceMap = new Map<string, Array<string>>();
  protectedResourceMap.set("https://graph.microsoft.com/v1.0/me", ["user.read"]);
  
  return {
    interactionType: InteractionType.Popup,
    protectedResourceMap,
    authRequest: {
      authority: 'https://login.microsoftonline.com/organizations',
    },

>     dynamicAuthRequest: (authRequest, httpReq) => {

Reviewer: ```suggestion
    authRequest: (msalService, httpReq) => {
```

      return {
        ...authRequest,

>         authority: `https://login.microsoftonline.com/${authRequest.account?.tenantId ?? 'organizations'}`

Reviewer: Instead of pull this off of the authRequest, you should use the get account APIs on the MSAL instance (either getActiveAccount, getAllAccounts[0], etc) depending on if your app sets the active account.

    });
  }
```

## Dynamic auth request

> By default the MsalGuard and the MsalInterceptor use the static properties you set in the config. Both can be configured with a method for the `authRequest`. This allows to dynamically change the parameters used for authentication.

Reviewer 3: ```suggestion
By default the MsalGuard and the MsalInterceptor use the static properties set in the config. Both can also be configured with a method for the `authRequest`, allowing the parameters used for authentication to be changed dynamically.
```

### MsalInterceptor - dynamic auth request (multi tenant tokens)

> If you're using `organizations` or `common` all the tokens will be requested for the users' home tenant, this might not result in the desired effect. If you have this application where you're invited as a guest the tokens will be from the wrong authority.

Reviewer 3: ```suggestion
If `organizations` or `common` is used as the tenant, all tokens will be requested for the users' home tenant. However, this may not be the desired outcome. If a user is invited as a guest, the tokens may be from the wrong authority.
```

> Setting the `authRequest` in the **MsalInterceptorConfig** to a method allows you to dynamically change the auth request, and for instance set the authority based on the home tenant of the account (when using guest users).

Reviewer 3: ```suggestion
Setting the `authRequest` in the **MsalInterceptorConfig** to a method allows you to dynamically change the auth request. For instance, you may set the authority based on the home tenant of the account when using guest users.
```

> You can change all properties but be sure to always extend the `originalAuthRequest` like you see below.

Reviewer 3: ```suggestion
Properties on `authRequest` may be changed, but should always extend the `originalAuthRequest` like below:
```
